pp-pkg/tsdb/db.go

```diff
@@ -0,0 +1,118 @@
+package tsdb
+
+import (
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/pp/go/relabeler/head/catalog"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"path/filepath"
+	"slices"
+)
+
+// PPBlocksToDelete returns a filter which decides time based and size based
+// retention from the options of the db.
+func PPBlocksToDelete(dir string, catalog *catalog.Catalog) func(db *tsdb.DB) tsdb.BlocksToDeleteFunc {
+	return func(db *tsdb.DB) tsdb.BlocksToDeleteFunc {
+		return func(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
+			return deletableBlocks(db, dir, catalog, blocks)
+		}
+	}
+}
+
+// deletableBlocks returns all currently loaded blocks past retention policy or already compacted into a new block.
+func deletableBlocks(db *tsdb.DB, dir string, catalog *catalog.Catalog, blocks []*tsdb.Block) map[ulid.ULID]struct{} {
+	deletable := make(map[ulid.ULID]struct{})
+
+	// Sort the blocks by time - newest to oldest (largest to smallest timestamp).
+	// This ensures that the retentions will remove the oldest  blocks.
+	slices.SortFunc(blocks, func(a, b *tsdb.Block) int {
+		switch {
+		case b.Meta().MaxTime < a.Meta().MaxTime:
+			return -1
+		case b.Meta().MaxTime > a.Meta().MaxTime:
+			return 1
+		default:
+			return 0
+		}
+	})
+
+	for _, block := range blocks {
+		if block.Meta().Compaction.Deletable {
+			deletable[block.Meta().ULID] = struct{}{}
+		}
+	}
+
+	for ulid := range BeyondTimeRetention(db, blocks) {
+		deletable[ulid] = struct{}{}
```

Comment:
For me it is better to pass set to function which will mutate it than copy mass of sets